// ng-thermal-print/esm2015/lib/drivers/WebPrintDriver.js
import { __awaiter } from "tslib";
import { BehaviorSubject } from 'rxjs';
import { PrintDriver } from "./PrintDriver";
import { StarWebPrintTrader } from '../js/StarWebPrintTrader.js';
export class WebPrintDriver extends PrintDriver {
    constructor(url, useSecure = false) {
        super();
        this.isConnected = new BehaviorSubject(false);
        this.useSecure = false;
        this.isStarPrinter = false;
        this.url = url;
        this.useSecure = useSecure;
    }
    connect() {
        const useSecure = (this.useSecure) ? 's' : '';
        this.trader = new StarWebPrintTrader({ url: `http${useSecure}://${this.url}/StarWebPRNT/SendMessage` });
        this.trader.onReceive = (response) => {
            this.isConnected.next(true);
        };
        this.trader.onError = (response) => {
            this.isConnected.next(false);
        };
        this.trader.sendMessage('');
    }
    write(data) {
        return __awaiter(this, void 0, void 0, function* () {
            this.trader.sendMessage({ request: data });
        });
    }
}
